Sources/Controllers/2.0/Settings/NavigationSettings/OARoadSpeedsViewController.mm

```diff
@@ -0,0 +1,272 @@
+{
+    [super viewDidLoad];
+    self.backButton.hidden = YES;
+    self.cancelButton.hidden = NO;
+    self.doneButton.hidden = NO;
+    self.tableView.delegate = self;
+    self.tableView.dataSource = self;
+    [self setupTableHeaderViewWithText:OALocalizedString(@"road_speeds_descr")];
+    [self setupView];
+}
+
+- (void) applyLocalization
+{
+    [super applyLocalization];
+    self.titleLabel.text = OALocalizedString(@"road_speeds");
+    [self.cancelButton setTitle:OALocalizedString(@"shared_string_cancel") forState:UIControlStateNormal];
+    [self.doneButton setTitle:OALocalizedString(@"shared_string_done") forState:UIControlStateNormal];
+}
+
+- (void) viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
+{
+    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
+    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
+        [self setupTableHeaderViewWithText:OALocalizedString(@"road_speeds_descr")];
+        [self.tableView reloadData];
+    } completion:nil];
+}
+
+- (void) commonInit
+{
+    [self generateData];
+}
+
+- (void) generateData
+{
+    auto router = [OARouteProvider getRouter:self.appMode];
+    _units = [OASpeedConstant toShortString:[_settings.speedSystem get:self.appMode]];
+    switch ([_settings.speedSystem get:self.appMode])
+    {
+        case MILES_PER_HOUR:
+            _ratio = 3600. / METERS_IN_ONE_MILE;
+            break;
+        case KILOMETERS_PER_HOUR:
+            _ratio = 3600. / METERS_IN_KILOMETER;
+            break;
+        case MINUTES_PER_KILOMETER:
+            _ratio = 3600. / METERS_IN_KILOMETER;
+            _units = OALocalizedString(@"units_kmh");
+            break;
+        case NAUTICALMILES_PER_HOUR:
+            _ratio = 3600. / METERS_IN_ONE_NAUTICALMILE;
+            break;
+        case MINUTES_PER_MILE:
+            _ratio = 3600. / METERS_IN_ONE_MILE;
+            _units = OALocalizedString(@"units_mph");
+            break;
+        case METERS_PER_SECOND:
+            _ratio = 1;
+            break;
+    }
+    CGFloat settingsMinSpeed = self.appMode.getMinSpeed;
+    CGFloat settingsMaxSpeed = self.appMode.getMaxSpeed;
+    CGFloat minSpeedValue = settingsMinSpeed > 0 ? settingsMinSpeed : router->getMinSpeed();
+    CGFloat maxSpeedValue = settingsMaxSpeed > 0 ? settingsMaxSpeed : router->getMaxSpeed();
+
+    _minValue = round(minSpeedValue * _ratio);
+    _maxValue = round(maxSpeedValue * _ratio);
+    _baseMinSpeed = MAX(1, floor(self.appMode.baseMinSpeed * _ratio));
+    _baseMaxSpeed = round(self.appMode.baseMaxSpeed * _ratio);
+}
+
+- (void) setupView
+{
+    NSMutableArray *tableData = [NSMutableArray array];
+    [tableData addObject:@{
+        @"type" : kCellTypeSpeed,
+        @"title" : OALocalizedString(@"logging_min_speed"),
+        @"value" : [NSString stringWithFormat:@"%ld %@", (long)_minValue, _units],
```

Comment:
Never cast NSInteger to long (I've removed such code in the file import too). NSInteger **is** a long.